### PR TITLE
NativeAOT: Enable building app extensions with NativeAOT

### PR DESCRIPTION
**Pull request by ivanpovazan**
### Description

This PR enables building app extensions with NativeAOT. 

App extensions are class libraries and to build them with NativeAOT we must not specify `CustomNativeMain=true`. If we do, ILC would expect that the input assembly has a managed Main as the module entry point.

Additionally, when building class libraries (with the absence of a managed Main entry point), our static reference from: 
https://github.com/xamarin/xamarin-macios/blob/2e5ef1eb1c14e185b29bbef091d227ed0c8cc875/runtime/nativeaot-bridge.m#L39

requires build-time symbol resolution. To avoid linking errors, we generate an empty `__managed__Main`
in the native bootstrapping code of the app extension (e.g., in `main.arm64.mm`).

### Testing

The unit tests have been introduced to test building app extensions with both Mono and NativeAOT.
Executing an iOS app TodayExtension built with NativeAOT has been verified manually on an actual device.

--- 
Fixes: https://github.com/xamarin/xamarin-macios/issues/20653

### CODE REVIEW

File: tools/dotnet-linker/LinkerConfiguration.cs [hunk 1]
```diff
@@ -236,6 +236,9 @@ public static LinkerConfiguration GetInstance (LinkContext context)
 				case "Optimize":
 					user_optimize_flags = value;
 					break;
+				case "OutputType":
+					Application.IsClassLibrary = string.Equals ("Library", value, StringComparison.OrdinalIgnoreCase);
+					break;
```
Comment (rolfbjarne) [inline tools/dotnet-linker/LinkerConfiguration.cs:241]: There's the existing IsAppExtension property you can use instead.

File: dotnet/targets/Xamarin.Shared.Sdk.props [hunk 2]
```diff
@@ -229,7 +229,12 @@
 		<!-- Disable our own assembly IL stripping logic, because ILC does that already -->
 		<EnableAssemblyILStripping>false</EnableAssemblyILStripping>
 
-		<!-- We're using our own native main function when using NativeAOT -->
-		<CustomNativeMain>true</CustomNativeMain>
+		<!--
+			We're using our own native main function when using NativeAOT.
+			This is true for both: managed executables and app extensions that can be class libraries.
```
Comment (rolfbjarne) [inline dotnet/targets/Xamarin.Shared.Sdk.props:234]: App extensions are always libraries:

```suggestion
			This is true for both: managed executables and app extensions (which are libraries).
```

File: tests/dotnet/UnitTests/ProjectTest.cs [hunk 3]
```diff
@@ -1063,46 +1063,67 @@ public void KillEverything ()
...
 
-		[TestCase (ApplePlatform.iOS)]
-		[TestCase (ApplePlatform.TVOS)]
-		[TestCase (ApplePlatform.MacOSX)]
+		[TestCase (ApplePlatform.iOS, "iossimulator-x64", false)]
+		[TestCase (ApplePlatform.iOS, "iossimulator-x64", true)]
+		[TestCase (ApplePlatform.TVOS, "tvossimulator-x64", false)]
+		[TestCase (ApplePlatform.TVOS, "tvossimulator-x64", true)]
+		[TestCase (ApplePlatform.MacOSX, "osx-x64", false)]
+		[TestCase (ApplePlatform.MacOSX, "osx-x64", true)]
```
Comment (rolfbjarne) [inline tests/dotnet/UnitTests/ProjectTest.cs:1106]: Does it work with universal apps?

```suggestion
		[TestCase (ApplePlatform.MacOSX, "osx-x64", false)]
		[TestCase (ApplePlatform.MacOSX, "osx-x64", true)]
		[TestCase (ApplePlatform.MacOSX, "osx-x64;osx-arm64", false)]
		[TestCase (ApplePlatform.MacOSX, "osx-x64;osx-arm64", true)]
```

File: tests/dotnet/UnitTests/ProjectTest.cs [hunk 4]
```diff
@@ -1063,46 +1063,67 @@ public void KillEverything ()
...
-		[TestCase (ApplePlatform.TVOS)]
-		[TestCase (ApplePlatform.MacOSX)]
+		[TestCase (ApplePlatform.iOS, "ios-arm64", false)]
+		[TestCase (ApplePlatform.iOS, "ios-arm64", true)]
+		[TestCase (ApplePlatform.iOS, "iossimulator-x64", false)]
+		[TestCase (ApplePlatform.iOS, "iossimulator-x64", true)]
+		[TestCase (ApplePlatform.TVOS, "tvossimulator-x64", false)]
+		[TestCase (ApplePlatform.TVOS, "tvossimulator-x64", true)]
+		[TestCase (ApplePlatform.MacOSX, "osx-x64", false)]
+		[TestCase (ApplePlatform.MacOSX, "osx-x64", true)]
```
Comment (rolfbjarne) [inline tests/dotnet/UnitTests/ProjectTest.cs:1073]: Does it work with universal apps?

```suggestion
		[TestCase (ApplePlatform.MacOSX, "osx-x64", false)]
		[TestCase (ApplePlatform.MacOSX, "osx-x64", true)]
		[TestCase (ApplePlatform.MacOSX, "osx-x64;osx-arm64", false)]
		[TestCase (ApplePlatform.MacOSX, "osx-x64;osx-arm64", true)]
```

**Review by rolfbjarne — CHANGES_REQUESTED** [review]

File: dotnet/targets/Xamarin.Shared.Sdk.props (hunk 2 above)
Comment (ivanpovazan) [reply dotnet/targets/Xamarin.Shared.Sdk.props]: Thanks, fixed.

File: tests/dotnet/UnitTests/ProjectTest.cs (hunk 3 above)
Comment (ivanpovazan) [reply tests/dotnet/UnitTests/ProjectTest.cs:1108]: It does not work with the universal apps.
In fact, I discovered that we have a more general issue when building universal apps with NativeAOT and opened: https://github.com/xamarin/xamarin-macios/issues/20903 about it.
For now, I commented out universal app extension tests with NativeAOT.

**Comment by ivanpovazan** [comment]
@mandel-macaque @dalexsoto 
could you please take a look?

Regarding Rolf's comment about including universal app extensions tests with NativeAOT I opened: https://github.com/xamarin/xamarin-macios/issues/20903 for tracking as we don't seem to properly support building regular universal apps with NativeAOT either.

**Review by dalexsoto — APPROVED** [review]

**Pull request merged**